File: internal/api-admin/route/check.go
// This file is part of the happyDomain (R) project.
// Copyright (c) 2020-2026 happyDomain
// Authors: Pierre-Olivier Mercier, et al.
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package route

import (
	"github.com/gin-gonic/gin"

	"git.happydns.org/happyDomain/internal/api-admin/controller"
)

func declareChecksRoutes(router *gin.RouterGroup, dep Dependencies) {
	cc := controller.NewCheckerController(dep.Checker)

	apiChecksRoutes := router.Group("/checks")

	apiChecksRoutes.GET("", cc.ListCheckers)

	apiCheckerRoutes := apiChecksRoutes.Group("/:cname")
	apiCheckerRoutes.Use(cc.CheckerHandler)

	apiCheckerRoutes.GET("", cc.GetCheckerStatus)
	//apiCheckerRoutes.POST("", tpc.ChangeCheckerStatus)

	apiCheckerRoutes.GET("/options", cc.GetCheckerOptions)
	apiCheckerRoutes.POST("/options", cc.AddCheckerOptions)
	apiCheckerRoutes.PUT("/options", cc.ChangeCheckerOptions)

	apiCheckerOptionsRoutes := apiCheckerRoutes.Group("/options/:optname")
	apiCheckerOptionsRoutes.Use(cc.CheckerOptionHandler)
	apiCheckerOptionsRoutes.GET("", cc.GetCheckerOption)
	apiCheckerOptionsRoutes.PUT("", cc.SetCheckerOption)
}

File: web/src/lib/model/check.ts
// This file is part of the happyDomain (R) project.
// Copyright (c) 2022-2026 happyDomain
// Authors: Pierre-Olivier Mercier, et al.
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import type {
    HappydnsCheckerAvailability,
    HappydnsCheckerOptionDocumentation,
    HappydnsCheckerOptionsDocumentation,
    HappydnsCheckerOptions,
    HappydnsCheckerResponse,
    HappydnsCheckerSchedule,
    HappydnsCheckResult,
    HappydnsCheckExecution,
} from "$lib/api-base/types.gen";

// Re-export auto-generated types with better names
export type CheckerAvailability = HappydnsCheckerAvailability;
export type CheckerInfo = HappydnsCheckerResponse;
export type CheckerList = { [key: string]: HappydnsCheckerResponse };
export type CheckerOptions = HappydnsCheckerOptions;
export type CheckerOptionsDocumentation = HappydnsCheckerOptionsDocumentation;
export type CheckerSchedule = HappydnsCheckerSchedule;
export type CheckResult = HappydnsCheckResult;
export type CheckExecution = HappydnsCheckExecution;

// Make 'id' required for CheckerOptionDocumentation
export interface CheckerOptionDocumentation extends Omit<HappydnsCheckerOptionDocumentation, "id"> {
    id: string;
}

// Enums for named access to numeric status/scope values
export enum CheckResultStatus {
    Unknown = 0,
    Crit = 1,
    Warn = 2,
    Info = 3,
    OK = 4,
}

export enum CheckScopeType {
    CheckScopeInstance = 0,
    CheckScopeUser = 1,
    CheckScopeDomain = 2,
    CheckScopeService = 3,
    CheckScopeOnDemand = 4,
}

export enum CheckExecutionStatus {
    CheckExecutionPending = 0,
    CheckExecutionRunning = 1,
    CheckExecutionCompleted = 2,
    CheckExecutionFailed = 3,
}

export interface AvailableCheck {
    checker_name: string;
    enabled: boolean;
    schedule?: CheckerSchedule;
    last_result?: CheckResult;
}
